frontends/parsers/p4/p4parser.ypp

```diff
@@ -395,6 +399,23 @@ fragment
           result->push_back(new IR::StringLiteral(@4, $4));
           result->push_back(new IR::StringLiteral(@6, $6));
           $$ = result; }
+
+      // P4Runtime annotations
+    | START_P4RT_TRANSLATION_ANNOTATION STRING_LITERAL "," p4rtControllerType
+        { auto* result = new IR::Vector<IR::Expression>();
+          result->push_back(new IR::StringLiteral(@2, $2));
+          result->push_back(new IR::DefaultExpression($4));
```

Comment:
@smolkaj as I mentioned in the other comment, I doubt you meant to use a `DefaultExpression` here (it corresponds to don't care or `_` in P4). I'm pretty sure this is connected to your test failures as well, because I just saw this error message in the Travis logs:
```
psa.p4(78): [--Werror=invalid] error: PortId_t: the second argument to @p4runtime_translation must be one of the basic types 'string' or 'bit<W>', but got _
```

Reply:
See my comments above, this is a hack.

Could it be that there is some type checking pass that overwrites my p4rtControllerType with a type `_`, because that's the type assigned to `DefaultExpression`? Maybe @mbudiu-vmw would know?

Reply:
Why not just encode the type into a string (a string literal) and decode it back when you need to look at it?
The real solution is to allow annotations also contain types, but for that we need to change the spec.

Reply:
I don't think we need a spec change here. To me, the issue is that the compiler's internal representation of parsed annotations simply isn't rich enough to capture what is needed for this annotation. One can imagine addressing this by specializing `Annotation` for the various kinds of annotations a compiler might need to handle.

Reply:
We should probably restructure `Annotation` that way anyways -- currently it contains an unparsed token vector, a parsed expression vector and a parsed kv-pair vector, even though the one of the latter two will always be empty.  It should instead be an abstract base class with subclasses `UnparsedAnnotation`, `ExpressionListAnnotation`, and `KVListAnnotation` -- the annotation parser would replace the unparsed annotation with the appropriate subclass.  Then adding a `TypeAnnotation` subclass would be easy.

Reply:
But this annotation has both an expression and a type.

Reply:
I think it would be a `P4rtTranslationAnnotation` class containing a string literal and a type.

Reply:
Having a `P4rtTranslationAnnotation` class seems nice.

In the meantime, I used a different encoding:
* `IR::StringLiteral()` encodes the type "string"
* `IR::Constant(W)` encodes the type "bit\<W\>"

It's still a hack, of course, but this seems to work.

Reply:
at the very least you should document the hack, and probably add a FIXME or TODO.

Reply:
+1 for documenting this with a comment

at least the new hack is not as hacky as the old hack IMO :)

Reply:
Sorry, I just forgot to push my documentation.
I also added a TODO